### PR TITLE
HDDS-7158. ldb cli command supports to scan container V3.

### PR DESCRIPTION
**Pull request by ChenSammi**
https://issues.apache.org/jira/browse/HDDS-7158

### CODE REVIEW

**Comment by kerneltime** [comment]
@duongnguyen0

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedSlice.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ */
+public class ManagedSlice extends Slice {
+
+  public ManagedSlice(byte[] var1) {
+    super(var1);
+  }
+
+  @Override
+  protected void finalize() throws Throwable {
+    ManagedRocksObjectMetrics.INSTANCE.increaseManagedObject();
```
Comment (duongnguyen0) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedSlice.java:34]: Let's make it an util in `ManagedRocksObjectMetrics`.
```
static void assertClosed(RocksMutableObject rocksObject) {
    ManagedRocksObjectMetrics.INSTANCE.increaseManagedObject();
    if (rocksObject.isOwningHandle()) {
      ManagedRocksObjectMetrics.INSTANCE.increaseLeakObject();
      LOG.warn("{} is not closed properly",
          rocksObject.getClass().getSimpleName());
    }
  }
```

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java [hunk 2]
```diff
@@ -100,13 +109,25 @@ private static List<Object> displayTable(ManagedRocksIterator iterator,
...
         StringBuilder result = new StringBuilder();
         if (withKey) {
           Object key = dbColumnFamilyDefinition.getKeyCodec()
               .fromPersistedFormat(iterator.get().key());
           Gson gson = new GsonBuilder().setPrettyPrinting().create();
-          result.append(gson.toJson(key));
+          if (schemaV3) {
+            int index =
+                DatanodeSchemaThreeDBDefinition.getContainerKeyPrefixLength();
+            String cid = ((String)key).substring(0, index);
```
Comment (duongnguyen0) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:124]:
```suggestion
            String cid = key.toString().substring(0, index);
```

File: hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java [hunk 3]
```diff
@@ -100,13 +109,25 @@ private static List<Object> displayTable(ManagedRocksIterator iterator,
...
+            String blockId = ((String)key).substring(index);
```
Comment (duongnguyen0) [inline hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:125]:
```suggestion
            String blockId = key.toString().substring(index);
```

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedSlice.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package org.apache.hadoop.hdds.utils.db.managed;
+
+import org.rocksdb.Slice;
+
+/**
+ * Managed Options.
```
Comment (duongnguyen0) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedSlice.java:24]:
```suggestion
 * Managed Slice.
```

**Review by duongnguyen0 — APPROVED** [review]
Thanks for the patch @ChenSammi. This looks good to me, only left a few nitpick comments.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedSlice.java (hunk 1 above)
Comment (ChenSammi) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedSlice.java:34]: @duongnguyen0 , thanks for the code review.  RocksMutableObject.isOwningHandle has a protected access privilege, can cannot be called in either ManagedRocksObjectMetrics or ManagedRocksObjectUtils(I think you assume this class).

Comment (duongnguyen0) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedSlice.java:34]: Sounds good, thanks for the clarification, Sammi.

**Review by guihecheng — APPROVED** [review]
LGTM+1
Thanks @ChenSammi !

**Comment by ChenSammi** [comment]
Thanks @duongnguyen0 and @guihecheng for the code review.

**Pull request merged**